File: poc/atproto/dds-reference/packages/client-ts/src/replay.ts
/**
 * Jetstream replay helpers for DDS project steps.
 *
 * A "project step" is a `(project_uri, phase)` pair. Replaying its history means: starting
 * from the project record's `createdAt`, walk the jetstream and yield every statement / vote /
 * group / metric / etc. whose `step` field points at this project + phase.
 *
 * These helpers are generic across consumers. They yield events; the caller decides what to do
 * with them (persist, render, etc.).
 *
 * Mirror of `dds_client.replay` (Python).
 */

import { Agent } from "@atproto/api";
import { subscribe, type DDSFeedEvent } from "./feed.js";

const DEFAULT_CATCHUP_GRACE_SECONDS = 2;
const DEFAULT_APPVIEW = "https://bsky.social";

export interface ResolveProjectCursorOptions {
  /** XRPC service to call `getRecord` against. Defaults to https://bsky.social. */
  service?: string;
  /** Optional pre-built agent (otherwise we use a bare fetch). */
  agent?: Agent;
}

/**
 * Resolve a project's `createdAt` to a unix-microsecond jetstream cursor. Returns null if the
 * URI is malformed, the record can't be fetched, or it has no parseable `createdAt`.
 */
export async function resolveProjectCursor(
  projectUri: string,
  options: ResolveProjectCursorOptions = {},
): Promise<number | null> {
  if (!projectUri.startsWith("at://")) return null;
  const parts = projectUri.slice(5).split("/");
  if (parts.length !== 3) return null;
  const [hostDid, collection, rkey] = parts as [string, string, string];

  let data: { value?: { createdAt?: unknown } } | null = null;
  try {
    if (options.agent) {
      const res = await options.agent.com.atproto.repo.getRecord({
        repo: hostDid,
        collection,
        rkey,
      });
      data = { value: res.data.value as { createdAt?: unknown } };
    } else {
      const service = options.service ?? DEFAULT_APPVIEW;
      const url = new URL("/xrpc/com.atproto.repo.getRecord", service);
      url.searchParams.set("repo", hostDid);
      url.searchParams.set("collection", collection);
      url.searchParams.set("rkey", rkey);
      const res = await fetch(url.toString());
      if (!res.ok) return null;
      data = (await res.json()) as typeof data;
    }
  } catch {
    return null;
  }

  const createdAt = data?.value?.createdAt;
  if (typeof createdAt !== "string") return null;
  const ms = Date.parse(createdAt);
  if (!Number.isFinite(ms)) return null;
  return ms * 1000;
}

/** True if a record's `step` field points at the given project + phase. */
export function stepMatches(
  record: Record<string, unknown> | null | undefined,
  projectUri: string,
  phase: string,
): boolean {
  if (!record) return false;
  const step = (record.step ?? {}) as Record<string, unknown>;
  const project = (step.project ?? {}) as Record<string, unknown>;
  return project.uri === projectUri && step.phase === phase;
}

export interface ReplayProjectStepOptions {
  projectUri: string;
  phase: string;
  /** Collections to follow. Defaults to all DDS collections. */
  collections?: string[];
  /** Skip cursor resolution and start from this explicit unix-microsecond timestamp. */
  cursorUs?: number;
  /** Seconds-of-grace before we declare "caught up to now". Defaults to 2. */
  catchupGraceSeconds?: number;
  /** PDS service for the `getRecord` call that resolves the project's createdAt. */
  service?: string;
  /** Optional pre-built agent (used both for `getRecord` and as the catch-up fetcher). */
  agent?: Agent;
  /** Optional logger; defaults to a no-op. */
  logger?: { info: (msg: string) => void; error: (err: unknown) => void };
}

/**
 * Replay jetstream from the project's createdAt onward, yielding events whose `step` matches.
 *
 * Stops as soon as jetstream's emitted event timestamps reach within `catchupGraceSeconds` of
 * wall-clock now. After that the caller should rely on `subscribe(...)` (live mode) to keep up.
 */
export async function* replayProjectStep(
  opts: ReplayProjectStepOptions,
): AsyncGenerator<DDSFeedEvent, void, undefined> {
  let cursor = opts.cursorUs;
  if (cursor == null) {
    const resolved = await resolveProjectCursor(opts.projectUri, {
      service: opts.service,
      agent: opts.agent,
    });
    if (resolved == null) return;
    cursor = resolved;
  }

  const grace = (opts.catchupGraceSeconds ?? DEFAULT_CATCHUP_GRACE_SECONDS) * 1000;
  const catchupThresholdUs = (Date.now() - grace) * 1000;

  const controller = new AbortController();

  try {
    for await (const ev of subscribe({
      collections: opts.collections,
      cursor,
      agent: opts.agent,
      logger: opts.logger,
      signal: controller.signal,
    })) {
      const eventUs = ev.timeMs * 1000;
      if (eventUs >= catchupThresholdUs) {
        controller.abort();
        break;
      }
      if (ev.source !== "jetstream") continue;
      if (!stepMatches(ev.record, opts.projectUri, opts.phase)) continue;
      yield ev;
    }
  } finally {
    controller.abort();
  }
}

File: poc/atproto/dds-reference/python/dds_client/dds_client/replay.py
"""Jetstream replay helpers.

`replay_from_cursor` is the generic primitive: start at a unix-µs cursor, yield jetstream
events for the given collections, stop when caught up to wall-clock now.

`replay_project_step` is a thin wrapper that resolves a project's `createdAt` and filters
to events matching `(project_uri, phase)`. Other consumers (e.g. an indexer service that
wants to backfill all of a project's records regardless of phase) can call
`replay_from_cursor` directly.
"""

from __future__ import annotations

import asyncio
import logging
import time
from datetime import datetime
from typing import AsyncIterator, Optional

import httpx

from .feed import DDSFeedEvent, JETSTREAM_DEFAULT, subscribe
from .identity import resolve_did_pds

log = logging.getLogger(__name__)

DEFAULT_CATCHUP_GRACE_SECONDS = 2.0
DEFAULT_APPVIEW = "https://bsky.social"


async def resolve_project_cursor(
    project_uri: str,
    *,
    service: str = DEFAULT_APPVIEW,
) -> Optional[int]:
    """Resolve a project's `createdAt` to a unix-microsecond jetstream cursor."""
    if not project_uri.startswith("at://"):
        return None
    parts = project_uri[len("at://"):].split("/")
    if len(parts) != 3:
        return None
    host_did, collection, rkey = parts

    base = (await resolve_did_pds(host_did)) or service

    async with httpx.AsyncClient(base_url=base, timeout=30.0) as http:
        try:
            resp = await http.get(
                "/xrpc/com.atproto.repo.getRecord",
                params={"repo": host_did, "collection": collection, "rkey": rkey},
            )
            resp.raise_for_status()
            data = resp.json()
        except Exception as exc:
            log.warning("resolve_project_cursor: getRecord(%s) failed: %s", project_uri, exc)
            return None

    created = (data.get("value") or {}).get("createdAt")
    if not isinstance(created, str):
        return None
    try:
        dt = datetime.fromisoformat(created.replace("Z", "+00:00"))
        return int(dt.timestamp() * 1_000_000)
    except Exception:
        return None


def step_matches(record: dict, project_uri: str, phase: Optional[str] = None) -> bool:
    """Return True if a record's `step` field points at the given project (and phase, if given).

    If `phase` is None, only the project URI is matched (any phase).
    """
    if not record:
        return False
    step = record.get("step") or {}
    project = step.get("project") or {}
    if project.get("uri") != project_uri:
        return False
    if phase is not None and step.get("phase") != phase:
        return False
    return True


async def replay_from_cursor(
    *,
    cursor_us: int,
    collections: Optional[list[str]] = None,
    dids: Optional[list[str]] = None,
    catchup_grace_seconds: float = DEFAULT_CATCHUP_GRACE_SECONDS,
    jetstream_endpoint: str = JETSTREAM_DEFAULT,
    logger=lambda msg: None,
) -> AsyncIterator[DDSFeedEvent]:
    """Generic: yield jetstream events from a unix-µs cursor until caught up to wall-clock now.

    The cutoff is `now - catchup_grace_seconds`. Once events arrive at or beyond that timestamp
    we close the subscription and return.

    The caller can do any filtering they need (project URI, target, author...). For the common
    "events for project step X" case, see `replay_project_step`.
    """
    catchup_threshold_us = int(time.time() * 1_000_000) - int(
        catchup_grace_seconds * 1_000_000
    )

    stop_event = asyncio.Event()

    async for ev in subscribe(
        collections=collections,
        dids=dids,
        cursor_us=cursor_us,
        jetstream_endpoint=jetstream_endpoint,
        stop_event=stop_event,
        logger=logger,
    ):
        event_us = int(ev.time_ms * 1000)
        if event_us >= catchup_threshold_us:
            stop_event.set()
            break
        if ev.source != "jetstream":
            continue
        yield ev


async def replay_project_step(
    *,
    project_uri: str,
    phase: Optional[str] = None,
    collections: Optional[list[str]] = None,
    cursor_us: Optional[int] = None,
    catchup_grace_seconds: float = DEFAULT_CATCHUP_GRACE_SECONDS,
    service: str = DEFAULT_APPVIEW,
    jetstream_endpoint: str = JETSTREAM_DEFAULT,
    logger=lambda msg: None,
) -> AsyncIterator[DDSFeedEvent]:
    """Replay jetstream from the project's createdAt, yielding events matching the project step.

    If `phase` is None, any phase for the project matches (useful for project-wide backfills).
    If `cursor_us` is not provided, we resolve it from the project record's createdAt.
    """
    if cursor_us is None:
        cursor_us = await resolve_project_cursor(project_uri, service=service)
        if cursor_us is None:
            return

    async for ev in replay_from_cursor(
        cursor_us=cursor_us,
        collections=collections,
        catchup_grace_seconds=catchup_grace_seconds,
        jetstream_endpoint=jetstream_endpoint,
        logger=logger,
    ):
        if not step_matches(ev.record, project_uri, phase):
            continue
        yield ev
